Hardware information implementation

amete opened the pull request:
This is still WIP but I think the basic idea is there. Therefore, it might be good to have a discussion before expanding on it.

Now the main interface has a `get_hardware_info()` method that is implemented in each plug-in to collect and return some hardware information. A basic implementation is done in `cpumon` to populate the chip vendor name as well as the number of processors that are defined. 

The collection is off by default and can be enabled w/ an additional argument, `--store-hw-info`, on the `prmon` side.  This results in the creating of a new section in the JSON file, named `HW`, and looks something like this at the moment:

```

  "Avg": {
    "nprocs": 1.0,
    "nthreads": 0.0,
    "pss": 122.0,
    "rchar": 2123.9583333333335,
    "read_bytes": 0.0,
    "rss": 548.0,
    "rx_bytes": 302.6041666666667,
    "rx_packets": 1.5625,
    "swap": 0.0,
    "tx_bytes": 647.1354166666666,
    "tx_packets": 2.6041666666666665,
    "vmem": 4372.0,
    "wchar": 0.0,
    "write_bytes": 0.0 
  },  
  "HW": {
    "cpu": {
      "model name": " Intel(R) Xeon(R) CPU E5-2667 v2 @ 3.30GHz",
      "nCPU": "32"
    }   
  },  
  "Max": {
    "nprocs": 1,
    "nthreads": 0,
    "pss": 122,
    "rchar": 8156,
    "read_bytes": 0,
    "rss": 548,
    "rx_bytes": 1162,
    "rx_packets": 6,
    "stime": 0,
    "swap": 0,
    "tx_bytes": 2485,
    "tx_packets": 10, 
    "utime": 0,
    "vmem": 4372,
    "wchar": 0,
    "write_bytes": 0,
    "wtime": 3
  }
}
```

I think we should have a discussion on what metrics we want to collect for in which plug-ins. The usual `/proc` exposes a number of things that we collect. I personally would like to stay away from using command-line utilities (especially those that require root privileges) but GPU information will possibly be an exception to that.

In any case, this is still a prototype but a decent starting point IMO.

Closes #114.

graeme-a-stewart commented on package/src/cpumon.cpp:
```diff
@@ -57,3 +57,45 @@ std::map<std::string, double> const cpumon::get_json_average_stats(
+    if (line.empty()) continue;
+    size_t splitIdx = line.find(":");
+    std::string val;
+    if (splitIdx != std::string::npos) val = line.substr(splitIdx + 1);
```
If we don't find the separator, shouldn't we also now skip the rest of the loop? e.g. put the metric test loop inside this `if`:

```
if (splitIdx != std::string::npos) {
  val = line.substr(splitIdx + 1);
  for (const auto& metric : metrics) {
    ...
  }
}
```

(or use `continue`)

amete replied:
That's correct, I reorganized this bit.

graeme-a-stewart commented on package/src/prmon.cpp:
```diff
@@ -260,6 +270,8 @@ int main(int argc, char* argv[]) {
         << "[--interval, -i TIME]     Seconds between samples (default "
         << default_interval << ")\n"
+        << "[--store-hw-info, -s]     Store hardware information (default "
+        << default_store_hw_info << ")\n"
```
This prints a `0` which isn't really nice. How about changing to
```
<< default_store_hw_info ? "true" : "false" << ...
```

amete replied:
Good point, updated.

graeme-a-stewart commented on package/src/memmon.cpp:
```diff
@@ -75,3 +75,9 @@ std::map<std::string, double> const memmon::get_json_average_stats(
+// Collect related hardware information
+std::map<std::string, std::map<std::string, std::string>> const memmon::get_hardware_info() {
+  std::map<std::string, std::map<std::string, std::string>> result{};
+  return result; 
```
It would be nice here to parse `/proc/meminfo` and get `MemTotal`

amete replied:
Good point, now added.

amete commented:
Hi @graeme-a-stewart,

Thanks a lot. I tidied things up a bit based on your comments. We now report the total available memory and I added some extra CPU information. The output now looks something like this:

```
{
  "Avg": {
    "nprocs": 1.0,
    "nthreads": 0.0,
    "pss": 190.0,
    "rchar": 523.9583333333334,
    "read_bytes": 0.0,
    "rss": 616.0,
    "rx_bytes": 1146.6145833333333,
    "rx_packets": 6.510416666666667,
    "swap": 0.0,
    "tx_bytes": 1371.3541666666667,
    "tx_packets": 7.03125,
    "vmem": 4432.0,
    "wchar": 0.0,
    "write_bytes": 0.0 
  },  
  "HW": {
    "cpu": {
      "model name": "Intel(R) Xeon(R) CPU E5-2667 v2 @ 3.30GHz",
      "nCPU": "32",
      "nCoresPerSocket": "8",
      "nSockets": "2",
      "nThreadsPerCore": "2" 
    },  
    "mem": {
      "MemTotal": "32723876 kB"
    }   
  },  
  "Max": {
    "nprocs": 1,
    "nthreads": 0,
    "pss": 190,
    "rchar": 2012,
    "read_bytes": 0,
    "rss": 616,
    "rx_bytes": 4403,
    "rx_packets": 25, 
    "stime": 0,
    "swap": 0,
    "tx_bytes": 5266,
    "tx_packets": 27, 
    "utime": 0,
    "vmem": 4432,
    "wchar": 0,
    "write_bytes": 0,
    "wtime": 3
  }
}
```

For the extra CPU information, I added some extra information as to how I compute them (essentially `nCPU = nSockets * nCoresPerSocket * nThreadsPerCore`). A check would be nice but a least on my machine `lscpu` reports consistent results:

```
CPU(s):                32
Thread(s) per core:    2
Core(s) per socket:    8
Socket(s):             2
```

Let me know how it looks.

graeme-a-stewart commented on package/src/cpumon.cpp:
```diff
@@ -57,3 +58,56 @@ std::map<std::string, double> const cpumon::get_json_average_stats(
+  result["cpu"]["nCPU"] = std::to_string(nCPU);
+  result["cpu"]["nSockets"] = std::to_string(nSockets);
+  result["cpu"]["nCoresPerSocket"] = std::to_string(nCores);
+  result["cpu"]["nThreadsPerCore"] = std::to_string(nThreads);
```
I don't like that these values com out as strings, that seems unnatural. So thinking about this maybe the underlying storage object should be a JSON entity, instead of the more rigid map of strings. That would allow for a more natural representation of this information.

If we were re-mapping post-facto then we'd have to somehow communicate the intrinsic type back to the main program, which seems complicated and not necessary if we went the JSON route. As this information is gathered only once then performance isn't a concern.

amete replied:
Done (see https://github.com/HSF/prmon/pull/116#issuecomment-638838785 for more details).

graeme-a-stewart commented on package/src/memmon.cpp:
```diff
@@ -75,3 +76,40 @@ std::map<std::string, double> const memmon::get_json_average_stats(
+      if (val.empty()) continue;
+      for (const auto& metric : metrics) {
+        if (line.size() >= metric.size() && line.compare(0, metric.size(), metric) == 0) {
+          result["mem"][metric] = std::regex_replace(val, std::regex("^\\s+|\\s+$"), "");
```
I would report this without the `kB` units in it (consistency with the way the rest of prmon works, Taking into account the above comment, it should be a JSON number, not a string.

amete replied:
Done.

amete commented:
Hi @graeme-a-stewart,

Thanks a lot. I updated the interface a bit. Now instead of returning a map, we pass the `json` object by reference and fill it inside the `get_hardware_info` method. This way we're more flexible in terms of types, so currently we have:

```
{
  "Avg": {
    "nprocs": 1.0,
    "nthreads": 0.0,
    "pss": 186.0,
    "rchar": 2123.9583333333335,
    "read_bytes": 0.0,
    "rss": 612.0,
    "rx_bytes": 557.03125,
    "rx_packets": 3.125,
    "swap": 0.0,
    "tx_bytes": 1105.46875,
    "tx_packets": 4.166666666666667,
    "vmem": 4432.0,
    "wchar": 0.0,
    "write_bytes": 0.0 
  },  
  "HW": {
    "cpu": {
      "model name": "Intel(R) Xeon(R) CPU E5-2667 v2 @ 3.30GHz",
      "nCPU": 32, 
      "nCoresPerSocket": 8,
      "nSockets": 2,
      "nThreadsPerCore": 2
    },  
    "mem": {
      "MemTotal": 32723876
    }   
  },  
  "Max": {
    "nprocs": 1,
    "nthreads": 0,
    "pss": 186,
    "rchar": 8156,
    "read_bytes": 0,
    "rss": 612,
    "rx_bytes": 2139,
    "rx_packets": 12, 
    "stime": 0,
    "swap": 0,
    "tx_bytes": 4245,
    "tx_packets": 16, 
    "utime": 0,
    "vmem": 4432,
    "wchar": 0,
    "write_bytes": 0,
    "wtime": 3
  }
}
```

Now what I don't like too much is that the `nlohmann/json.hpp` is included in the interface becase forward declaring this guy is a bit awkward. The library needs to be built w/ a special flag `-DJSON_MultipleHeaders=ON` which also gives an additional `fwd` header but I couldn't get it to work quite nicely. I'm not sure if this is a showstopper for you but at least it seems to be working.

Let me know what you think.

graeme-a-stewart commented on package/src/Imonitor.h:
```diff
@@ -25,6 +27,7 @@ class Imonitor {
   virtual std::map<std::string, unsigned long long> const get_json_total_stats() = 0;
   virtual std::map<std::string, double> const get_json_average_stats(unsigned long long elapsed_clock_ticks) = 0;
 
+  virtual void const get_hardware_info(nlohmann::json& j) = 0;
```
(and for all the concrete implmentations as well)

amete replied:
Indeed, `j` is fairly obscure, switched to `hw_json` throughout.

amete commented:
I just removed the WIP, I think this is ready to go whenever you're happy w/ it @graeme-a-stewart. Thanks a lot 😄
